### ISSUE TITLE
Fault in memalign leads to memory corruption

### ISSUE THREAD

**Issue by stephen-williams-capgemini** [seL4/musllibc#17]
This issue was previously the subject a [discussion on the devel mailing list](https://lists.sel4.systems/hyperkitty/list/devel@sel4.systems/message/2K2S7OOZJCWD3Z22XGS36OIXD7HRXKNC/). The conclusion of the discussion was the detection of a fault in the implementation of memalign that results in memory corruption.

**Fault Summary:**

The correctness of the memory allocation bookkeeping system relies upon the constraint that the minimum size of a memory 'chunk' is 4xsizeof(size_t). If this constraint is broken then bad things happen and the bookkeeping system becomes corrupted, specifically:

1. Arithmetic wrap-around of x occurs in the routines bin_index and bin_index_up within malloc.c. This can lead to chunks below the minimum size limit to be considered to be large unallocated chunks of memory. Subsequent allocation of these unallocated chunks (considered to be large but in reality tiny) allows previously allocated chunks to be re-used / overwritten.

2. The 'next' and 'prev' pointers held in an unallocated chunk (used to maintained a doubly linked list of unallocated chunks) that is below the minimum size limit may be overlayed with the bookkeeping of the following chunk.

The malloc routine enforces this minimum chunk size limit, however the code of the __memalign routine within memalign.c can break this minimum size constraint and therefore lead to corruption of the bookkeeping. 

The __memalign routine works by malloc'ing sufficient memory to ensure the requested amount of memory is available, at the requested alignment, somewhere within the malloc'ed region. This means that there may be some unused memory allocated before the start of the aligned memory area. This is handled by splitting the chunk allocated by malloc into two chunks, a chunk of memory prior to the start of the aligned memory followed by a chunk that starts at the requested alignment. __memalign then calls 'free' on the first chunk which wasn't required. So far so good, however __memalign fails to enforce the minimum chunk size constraint on either of the two split chunks.

In our case the first chunk (the one to be freed) was only 16 bytes whilst 4xsizeof(size_t) is 32 bytes, thereby breaking the minimum chunk size constraint and so led to the detected corruption.

**Proposed fix**

The following patch to memalign which has been tested and has been shown to resolve the identified weakness:

``` c
diff --git a/src/malloc/memalign.c b/src/malloc/memalign.c
index 006bd21c..c98963f0 100644
--- a/src/malloc/memalign.c
+++ b/src/malloc/memalign.c
@@ -21,6 +21,9 @@ void *__memalign(size_t align, size_t len)
                errno = ENOMEM;
                return NULL;
        }
+       else if (len < 4*sizeof(size_t))
+               /* Ensure the length of returned chunk meets the minimum limit. */
+               len = 4*sizeof(size_t);
 
        if (align <= 4*sizeof(size_t)) {
                if (!(mem = malloc(len)))
@@ -50,7 +53,29 @@ void *__memalign(size_t align, size_t len)
        ((size_t *)new)[-1] = header&7 | end-new;
        ((size_t *)end)[-2] = footer&7 | end-new;
 
-       free(mem);
+       if (new-mem >= 4*sizeof(size_t))
+               free(mem);
+       else {
+               /* The size of the region before 'new' is too small to be handled as a chunk
+                * so we cannot call 'free' on it. Instead we either discard the memory or
+                * transfer ownership of it to the previous chunk */
+               if (!(((size_t *)mem)[-2] & -8)) {
+                       /* mem->psize has no length, i.e. 'mem' is the first chunk in this mapped
+                        * region. In this case we simply discard the memory prior to 'new' by
+                        * making 'new' the first chunk of the mapped region. To do this we set
+                        * the length of new->psize to 0 with the 'in use' flag set, which equates
+                        * to simply setting its value to 1. */
+                       ((size_t *)new)[-2] = 1;
+               }
+               else {
+                       /* There is a previous chunk, assign ownership of the region before 'new'
+                        * to this previous chunk by increasing it's length. */
+                       unsigned char *pre = mem - (((size_t *)mem)[-2] & -8);
+                       ((size_t *)pre)[-1] += new-mem;
+                       ((size_t *)new)[-2] = ((size_t *)pre)[-1];
+               }
+       }
+
        return new;
 }
```

**Comment by axel-h** [comment]
Thanks. Could you create a PR with this patch?

**Comment by stephen-williams-capgemini** [comment]
https://github.com/seL4/musllibc/pull/18 raised as requested.

**Comment by stephen-williams-capgemini** [comment]
Would it be helpful to schedule a zoom / Teams / etc discussion so I can walk someone through the underlying fault and the supplied patch?

**Comment by kent-mcleod** [comment]
> Would it be helpful to schedule a zoom / Teams / etc discussion so I can walk someone through the underlying fault and the supplied patch?

Would you be able to provide the memalign arguments that was causing malloc data structures to be corrupted?

**Comment by kent-mcleod** [comment]
> The correctness of the memory allocation bookkeeping system relies upon the constraint that the minimum size of a memory 'chunk' is 4xsizeof(size_t). If this constraint is broken then bad things happen and the bookkeeping system becomes corrupted, specifically:

The part that I'm still struggling to convince myself is after malloc successfully returns the allocation, whether musl is correct to assume that if `new != mem` then a chunk associated with `new` is non-overlapping with a chunk associated with `mem`.  The current version of that function still appears to assume this and doesn't appear to have otherwise changed the way it calls malloc or the malloc book keeping structures: https://git.musl-libc.org/cgit/musl/tree/src/malloc/oldmalloc/aligned_alloc.c#n31 

It would seem that your breaking examples would still break on mainline libmusl that is using `oldmalloc`?

**Comment by stephen-williams-capgemini** [comment]
> > Would it be helpful to schedule a zoom / Teams / etc discussion so I can walk someone through the underlying fault and the supplied patch?
> 
> Would you be able to provide the memalign arguments that was causing malloc data structures to be corrupted?

Unfortunately it's not possible to provide arguments that are guaranteed to trigger the issue. The trigger conditions are a combination of the arguments to memalign (which we are in control of) as well as the alignment of the address returned by malloc (which we cannot control).

One simple set of conditions that were able to trigger the issue is shown in the original discussion [here](https://lists.sel4.systems/hyperkitty/list/devel@sel4.systems/message/4UG5D2VBXU2A7FMWYU45NDTCXFLY6WYU/). In this case it was as simple as a test program making two consecutive calls to "memalign(0x40, 0x1000)" which both returned the same address.

What I can say is that the probability of triggering the problem increases as the requested alignment decreases.

**Comment by stephen-williams-capgemini** [comment]
> > The correctness of the memory allocation bookkeeping system relies upon the constraint that the minimum size of a memory 'chunk' is 4xsizeof(size_t). If this constraint is broken then bad things happen and the bookkeeping system becomes corrupted, specifically:
> 
> The part that I'm still struggling to convince myself is after malloc successfully returns the allocation, whether musl is correct to assume that if `new != mem` then a chunk associated with `new` is non-overlapping with a chunk associated with `mem`. The current version of that function still appears to assume this and doesn't appear to have otherwise changed the way it calls malloc or the malloc book keeping structures: https://git.musl-libc.org/cgit/musl/tree/src/malloc/oldmalloc/aligned_alloc.c#n31
> 
> It would seem that your breaking examples would still break on mainline libmusl that is using `oldmalloc`?

Agreed, even though the code within `oldmalloc` branch of libmusl has moved on from the fork that exists within seL4 the same underlying bug seems to be present, i.e. as you've suggested it assumes that when `new != mem` the chunks are not overlapping / meet the minimum chunk size limit.

On that basis I will separately raise this bug with the libmusl project.

Given the differences in the code bases that exist between mainline libmusl and the seL4 fork I'm assuming that you still wish to proceed with the patch I have provided, i.e. the fix for mainline libmusl would not be compatible with the seL4 fork.

**Comment by stephen-williams-capgemini** [comment]
I've raised a bug report with the musl project (see [thread starting here](https://www.openwall.com/lists/musl/2022/05/03/1l)) which has provided interesting information.

It seems that my patch is working around / protecting against a case that is not supposed to be possible. Apparently the addresses returned by malloc are expected to be a multiple of SIZE_ALIGN (equal to `4 x sizeof(size_t)`). The musl code has always expected this invariant to hold, and the bug is only triggered when this is not the case.

So it would seem we really need to determine why the address returned by malloc is not as expected. Are you able to answer the questions Rich (from the musl mailing list) has asked about PAGE_SIZE and libc.page_size?

**Comment by axel-h** [comment]
Link fix: https://www.openwall.com/lists/musl/2022/05/03/1

**Comment by kent-mcleod** [comment]
@stephen-williams-capgemini, I'm pretty sure that the issue is caused by unaligned morecore regions in CAmkES component environments (https://github.com/seL4/camkes-tool/pull/109).  With this change, multiple calls to `memalign(0x40, 0x1000)` stop returning overlapping results in a CAmkES test app I was using.

Rich's [comment](https://www.openwall.com/lists/musl/2022/05/03/4) about brk alignment is what helped track it down.

Our version of musl's `malloc` calls `expand_heap()` for more memory regions for allocations.  It expects that the return range is 4k aligned which wasn't being obeyed by the mock syscalls implementing `brk` and `mmap`.

I added some asserts to the mmap implementation to catch any alignment issues in the future: https://github.com/seL4/seL4_libs/pull/63

**Comment by kent-mcleod** [comment]
> I'm assuming that the https://github.com/seL4/musllibc/pull/18 will now be rejected as no longer required.

That's most likely correct.

Thanks for chasing this down and for all the time you spent diagnosing it!
